src/provider/store/Store.ts

```diff
@@ -1,122 +1,92 @@
+    private _currentState: S;
+    private _listeners: Listener<S>[] = [];
+
+    public readonly onAction: Signal<[A], Promise<void>, Promise<void>> = new Signal(Aggregators.AWAIT_VOID);
 
-        this._database = database;
-        this._actionHandlerMap = actionHandlerMap;
+    constructor(initialState: S) {
+        super();
+        this._currentState = initialState;
     }
 
-    protected async init(): Promise<void> {
-        await this._database.initialized;
-        this._store = createStore<RootState, RootAction, {}, {}>(this.reduce.bind(this), await this.getInitialState(), this.createEnhancer());
+    public get state(): S {
+        return this._currentState;
     }
 
-    public get state(): RootState {
-        return this._store.getState() as RootState;
+    public getState(): S {
+        return this._currentState;
     }
 
-    public async dispatch(action: RootAction): Promise<void> {
-        if (action instanceof CustomAction) {
+    protected async init(): Promise<void> {}
+
+    public async dispatch(action: A): Promise<void> {
+        if (action instanceof AsyncAction) {
             // Action has custom dispatch logic
             await action.dispatch(this);
-        } else {
-            // Pass straight through to redux store
-            await this._store.dispatch({...action});
         }
+        this.reduceAndEmit(action);
     }
 
-    public watchForChange<T>(getObject: (state: RootState) => T, observer: StoreChangeObserver<T>): Unsubscribe {
-        const watcher = this.watch<T>(() => this.state, getObject);
-        return this._store.subscribe(watcher(observer));
-    }
+    public subscribe(listener: Listener<S>) {
+        this._listeners.push(listener);
 
-    private watch<T>(getState: () => any, getObject: (state: RootState) => T): (observer: StoreChangeObserver<T>) => () => void {
-        let currentValue = getObject(getState());
-        return function w(observer: StoreChangeObserver<T>) {
-            return function () {
-                const newValue: T = getObject(getState());
-                if (currentValue !== newValue) {
-                    const oldValue = currentValue;
-                    currentValue = newValue;
-                    observer(oldValue, newValue);
-                }
-            };
+        return () => {
+            const index: number = this._listeners.indexOf(listener);
+            if (index >= 0) {
+                this._listeners.splice(index, 1);
+            }
         };
     }
 
-    private reduce<T extends Action>(state: RootState | undefined, action: ActionOf<T>): RootState {
-        const handler: ActionHandler<T> | undefined = this._actionHandlerMap[action.type] as ActionHandler<T>;
-
-        if (handler) {
-            return handler(state!, action);
-        } else {
-            // No handler registered for this action - action does not modify the store's state
-            return state!;
-        }
+    protected setState(state: S): void {
```

Comment:
Why state and getState?

Reply:
getState is used by react-redux, for the rest of our components we want to use this._store.state.  syntactic sugar :)

Reply:
Suggest make `getState` private (that shouldn't stop it working inside react-redux stuff), and have a one-line comment on why it's needed

Reply:
Think a one-line comment is still needed to say why we have this, and would move `getState` right to the very bottom - like, since we dont' want to use it ourselves, if we could make it super-extra-private, we would 😛. Think also get rid of it's use in `reduce` and use `this.state` or `this._currentState` instead